### PR TITLE
support added for ingress.kubernetes.io/force-ssl-redirect

### LINKED ISSUE #63: Support ingress.kubernetes.io/force-ssl-redirect
Add support for `ingress.kubernetes.io/force-ssl-redirect` per Ingress which will require creating a redirect route in the http version listener. This will need LDS to become a proper cache.

### PR DESCRIPTION
**Pull request by sch00lb0y**
close #63 
ingress.kubernetes.io/force-ssl-redirect annotation added for redirecting http to https.
Signed-off-by: பாலாஜி <rbalajis25@gmail.com>

### CODE REVIEW

**Comment by davecheney** [comment]
Thank you for this PR. This will need some tests before I can review it. Thank you.

File: internal/contour/listener.go [hunk 1]
```diff
@@ -194,6 +194,19 @@ func (lc *ListenerCache) httpsAddress() string {
 	return DEFAULT_HTTPS_LISTENER_ADDRESS
 }
 
+func (lc *listenerCache) HostHasTls(host string) bool {
+	for _, listener := range lc.Values() {
```
Comment (davecheney) [inline internal/contour/listener.go]: Please don't use lc.Values for this. Any information that you get out of lc.Values will be present in the other caches maintained by the listenerCache.

File: internal/contour/virtualhost.go [hunk 2]
```diff
@@ -145,3 +150,10 @@ func clusteraction(cluster string) *v2.Route_Route {
 		},
 	}
 }
+
+func sslRedirect(i *v1beta1.Ingress) bool {
+	if i.Annotations["kubenetes.io/ingress.ssl-redirect"] == "false" {
```
Comment (davecheney) [inline internal/contour/virtualhost.go]: this can be simplified to

   return i.Annotations["kubenetes.io/ingress.ssl-redirect"] == "true"

**Comment by sch00lb0y** [comment]
@davecheney  I have made changes. please review :+1:

File: internal/contour/listener.go [hunk 3]
```diff
@@ -194,6 +194,19 @@ func (lc *ListenerCache) httpsAddress() string {
 	return DEFAULT_HTTPS_LISTENER_ADDRESS
 }
 
+func (lc *listenerCache) HostHasTls(host string) bool {
```
Comment (davecheney) [inline internal/contour/listener.go]: You cannot use values here without taking the lock. 

Please discuss  your design before sending code, this is the second time I have asked.

File: internal/contour/virtualhost_test.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+package contour
+
+import (
+	"testing"
+
+	"k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func TestSslRedirect(t *testing.T) {
```
Comment (davecheney) [inline internal/contour/virtualhost_test.go]: This is a good test, but does not test the right thing. Then name of the annotation is
`ingress.kubernetes.io/force-ssl-redirect`

**Pull request closed without merging**